trackintel/preprocessing/triplegs.py

```diff
@@ -89,283 +86,204 @@ def generate_trips(stps_input, tpls_input, gap_threshold=15, print_progress=Fals
 
     tpls["type"] = "tripleg"
-    stps["type"] = "staypoint"
+    spts["type"] = "staypoint"
 
     # create table with relevant information from triplegs and staypoints.
-    stps_tpls = stps[["started_at", "finished_at", "user_id", "type", "activity"]].append(
-        tpls[["started_at", "finished_at", "user_id", "type"]]
+    spts_tpls = pd.concat(
+        [
+            spts[["started_at", "finished_at", "user_id", "type", "activity"]],
+            tpls[["started_at", "finished_at", "user_id", "type"]],
+        ]
     )
+    # transform nan to bool
+    spts_tpls["activity"].fillna(False, inplace=True)
 
     # create ID field from index
-    stps_tpls["id"] = stps_tpls.index
+    spts_tpls["id"] = spts_tpls.index
+
+    spts_tpls.sort_values(by=["user_id", "started_at"], inplace=True)
+    spts_tpls.reset_index(inplace=True, drop=True)
+
+    # delete intermediate activities
+    # intermediate activities are activities that are surrounded by other activities
+    # therefore they are neither at the end, at the start, nor within a trip
+    first_activities, inter_activities, last_activities = _get_activity_masks(spts_tpls)
+    spts_tpls["first_activity"] = first_activities
+    spts_tpls["last_activity"] = last_activities  # trip may start here
+    spts_tpls = spts_tpls[~inter_activities]  # no trip interaction
+    spts_tpls["started_at_next"] = spts_tpls["started_at"].shift(-1)
+
+    # conditions for new trip
+    # start new trip if the user changes
+    condition_new_user = spts_tpls["user_id"] != spts_tpls["user_id"].shift(1)
+
+    # start new trip if there is a new activity
+    condition_new_activity = spts_tpls["last_activity"]
+
+    # gap conditions
+    # start new trip after a gap
+    gap = (spts_tpls["started_at_next"] - spts_tpls["finished_at"]) > gap_threshold
+    condition_time_gap = gap.shift(1, fill_value=False)  # trip starts on next entry
+
+    cond_all = condition_new_user | condition_new_activity | condition_time_gap
+    spts_tpls["new_trip"] = cond_all
+
+    # assign an incrementing id to all triplegs that start a trip
+    spts_tpls.loc[cond_all, "trip_id"] = np.arange(cond_all.sum())
+    spts_tpls["trip_id"].fillna(method="ffill", inplace=True)
+
+    # exclude activities to aggregate trips together.
+    # activity can be thought of as the same level as trips.
+    spts_tpls_no_act = spts_tpls[~spts_tpls["activity"]]
+    spts_tpls_only_act = spts_tpls[spts_tpls["activity"]]
+
+    trips_grouper = spts_tpls_no_act.groupby("trip_id")
+    trips = trips_grouper.agg({"user_id": "mean", "started_at": min, "finished_at": max, "type": list, "id": list})
+
+    def _seperate_ids(row):
+        """Split aggregated ids into staypoints and triplegs arrays."""
+        row_type = np.array(row["type"])
+        row_id = np.array(row["id"])
+        t = row_type == "tripleg"
+        tpls_ids = row_id[t]
+        spts_ids = row_id[np.logical_not(t)]
```

Comment:
```suggestion
        spts_ids = row_id[~t]
```

Reply:
The `np.logical_not(t)` is here for singletons aka `True`. If we would use the bitwise operator `~` for those we get the value -2 (And -1 for False). I guess that is because of some type-casting in python.
We get singletons if there is only one entry in the list.

Reply:
I doubled checked you were right, that was an legacy issue from aggregating directly to numpy arrays.